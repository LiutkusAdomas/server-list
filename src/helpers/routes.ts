export const ROUTES = {
    SERVER_LIST: '/server-list',
    LOGIN: '/login',
    LOGOUT: '/logout'
};