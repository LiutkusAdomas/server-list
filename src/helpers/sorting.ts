import { Server } from "../model/Server.type";

export const ASCENDING = 'ascending';
export const DESCENDING = 'descending';

export const sortData = (data: Server[], sortConfig: { key: string; direction: string }) => {
    return [...data].sort((a, b) => {
        if (a[sortConfig.key as keyof Server] < b[sortConfig.key as keyof Server]) {
            return sortConfig.direction === ASCENDING ? -1 : 1;
        }
        if (a[sortConfig.key as keyof Server] > b[sortConfig.key as keyof Server]) {
            return sortConfig.direction === ASCENDING ? 1 : -1;
        }
        return 0;
    });
};
