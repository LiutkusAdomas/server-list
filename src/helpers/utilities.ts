import { Server } from "../model/Server.type";
import { ServerResponse } from "../model/ServerResponse.type";
import { v4 } from 'uuid';
import latvia from '../assets/latvia.svg';
import japan from '../assets/japan.svg';
import lithuania from '../assets/lithuania.svg';
import singapore from '../assets/singapore.svg';
import germany from '../assets/germany.svg';
import uk from '../assets/uk.svg';
import us from '../assets/us.svg';

export const transformResponse = (data: ServerResponse[]): Server[] => {
    return data.map(item => ({
        id: v4(),
        name: item.name,
        distance: item.distance,
        flag: getFlag(item.name)
    }));
}

export const validateInput = (value: string): string | undefined => {
    if (value.length < 6) {
        return 'This field must be longer than 5 characters';
    }
    else if (value.length > 20) {
        return 'This field must be shorter than 20 characters';
    }
    else return undefined;
}

const getFlag = (name: string): string => {
    if (name.toLowerCase().includes('latvia'))
        return latvia;
    else if (name.toLowerCase().includes('japan'))
        return japan;
    else if (name.toLowerCase().includes('lithuania'))
        return lithuania;
    else if (name.toLowerCase().includes('singapore'))
        return singapore;
    else if (name.toLowerCase().includes('germany'))
        return germany;
    else if (name.toLowerCase().includes('united kingdom'))
        return uk;
    else if (name.toLowerCase().includes('united states'))
        return us;
    else
        return '';
}