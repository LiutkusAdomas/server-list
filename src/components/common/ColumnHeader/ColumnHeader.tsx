import up from '../../../assets/up.svg';
import down from '../../../assets/down.svg';
import { ASCENDING } from '../../../helpers/sorting';

type ColumnHeaderPropsType = {
  label: string;
  sortKey: string;
  sortConfig: { key: string; direction: string };
  requestSort: (key: string) => void;
};

export const ColumnHeader: React.FC<ColumnHeaderPropsType> = ({
  label,
  sortKey,
  sortConfig,
  requestSort,
}) => {
  return (
    <button type="button" onClick={() => requestSort(sortKey)} className="text-gray-800">
      {label}
      {sortConfig.key === sortKey && (
        <img
          className="h-6 w-6 inline-block -mt-1"
          alt="Direction"
          src={sortConfig.direction === ASCENDING ? up : down}
        />
      )}
    </button>
  );
};
