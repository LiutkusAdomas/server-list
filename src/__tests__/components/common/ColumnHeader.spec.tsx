import { render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { ColumnHeader } from '../../../components/common/ColumnHeader/ColumnHeader';
import { ASCENDING, DESCENDING } from '../../../helpers/sorting';

describe('Column Header component', () => {
  it('should render the Column Header with the visual props and ascending arrow', () => {
    const { asFragment } = render(
      <ColumnHeader
        label="Test"
        sortKey="test"
        sortConfig={{ key: 'test', direction: ASCENDING }}
        requestSort={vi.fn()}
      />
    );
    expect(asFragment).toMatchSnapshot();
  });
  it('should render the Column Header with the visual props and descending arrow', () => {
    const { asFragment } = render(
      <ColumnHeader
        label="Test"
        sortKey="test"
        sortConfig={{ key: 'test', direction: DESCENDING }}
        requestSort={vi.fn()}
      />
    );
    expect(asFragment).toMatchSnapshot();
  });
});
