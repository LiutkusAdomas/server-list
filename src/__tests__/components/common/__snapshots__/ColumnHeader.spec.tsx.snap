// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Column Header component > should render the Column Header with the visual props and ascending arrow 1`] = `[Function]`;

exports[`Column Header component > should render the Column Header with the visual props and descending arrow 1`] = `[Function]`;
